src/main.rs

```diff
@@ -1,28 +1,131 @@
-extern crate cargo;
-extern crate rustc_serialize;
-extern crate docopt;
+use std::path::{Path, PathBuf};
+use std::process;
 
-#[macro_use]
-mod common;
-mod resolve;
-mod scan;
+use anyhow::{anyhow, Result};
+use cargo_metadata::{Dependency, Metadata, MetadataCommand, Package};
+use structopt::StructOpt;
 
-use cargo::util::Config;
-use docopt::Docopt;
-use rustc_serialize::json;
+mod types;
 
-const USAGE: &'static str = "
-Usage: srclib-rust scan [--repo=<repo>] [--subdir=<subdir>]
-";
+use crate::types::*;
+
+static MANIFEST_FILE: &str = "Cargo.toml";
+static UNIT_TYPE: &str = "RustCargoPackage";
+static USAGE: &str = "srclib-rust scan [--repo=<repo>] [--subdir=<subdir>]";
+
+#[derive(Debug, StructOpt)]
+struct ParsedArgs {
+    #[structopt(long, default_value)]
+    repo: String,
+    #[structopt(long, default_value = ".")]
+    pub subdir: PathBuf,
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "srclib-rust", usage = USAGE)]
+enum CmdRoot {
+    Scan(ParsedArgs),
+}
+
+fn get_manifest_path<P: AsRef<Path>>(root: &P) -> PathBuf {
+    root.as_ref().join(MANIFEST_FILE)
+}
+
+fn parse_args() -> ParsedArgs {
+    match CmdRoot::from_args() {
+        CmdRoot::Scan(parsed) => parsed,
+    }
+}
+
+fn parse_cargo_meta(args: &ParsedArgs) -> Result<Metadata> {
+    MetadataCommand::new()
+        .manifest_path(get_manifest_path(&args.subdir)) // use subdir arg
+        .no_deps() // Don't get dependency meta
+        .other_options(&["--offline".into()]) // don't make network calls
+        .exec()
+        .map_err(|e| e.into())
+}
+
+fn xform_meta(meta: Metadata, args: &ParsedArgs) -> Result<Vec<SourceUnit>> {
+    let this_package: &Package = get_this_package(&meta);
+    let src_unit = SourceUnitBuilder::default()
+        .name(this_package.name.to_owned())
+        .unit_type(UNIT_TYPE.to_owned())
+        .repo(Some(args.repo.clone()))
+        .files(get_source_files(&args.subdir)?)
+        .deps(Some(get_direct_deps(&this_package, args)?))
+        .data(this_package.license.to_owned().map(Into::<License>::into))
```

Comment:
I think it might be more idiomatic to add a constructor to `License` and use `.map(License::new)` here? (and kill the From<String>)

Reply:
In my experience, when type only serves any purpose as a step in a data-xform pipeline, it's much better to use `from`.  Also, since the `License` is just a wrapping type, it's common to have a `From` impl.